data-prepper-plugins/mongodb/src/main/java/org/opensearch/dataprepper/plugins/mongo/export/ExportScheduler.java

```diff
@@ -0,0 +1,116 @@
+
+    public ExportScheduler(final EnhancedSourceCoordinator enhancedSourceCoordinator,
+                           final MongoDBExportPartitionSupplier mongoDBExportPartitionSupplier,
+                           final PluginMetrics pluginMetrics) {
+        this.enhancedSourceCoordinator = enhancedSourceCoordinator;
+        this.mongoDBExportPartitionSupplier = mongoDBExportPartitionSupplier;
+        this.pluginMetrics = pluginMetrics;
+
+        executor = Executors.newCachedThreadPool();
+
+        exportJobSuccessCounter = pluginMetrics.counter(EXPORT_JOB_SUCCESS_COUNT);
+        exportJobFailureCounter = pluginMetrics.counter(EXPORT_JOB_FAILURE_COUNT);
+        exportPartitionTotalCounter = pluginMetrics.counter(EXPORT_PARTITION_QUERY_TOTAL_COUNT);
+        exportRecordsTotalCounter = pluginMetrics.counter(EXPORT_RECORDS_TOTAL_COUNT);
+    }
+
+    @Override
+    public void run() {
+        LOG.info("Start running Export Scheduler");
+        while (!Thread.currentThread().isInterrupted()) {
+            try {
+                final Optional<EnhancedSourcePartition> sourcePartition = enhancedSourceCoordinator.acquireAvailablePartition(ExportPartition.PARTITION_TYPE);
+                if (sourcePartition.isPresent()) {
+                    final ExportPartition exportPartition = (ExportPartition) sourcePartition.get();
+                    LOG.info("Acquired an export partition: " + exportPartition.getPartitionKey());
+
+                    final List<PartitionIdentifier> partitionIdentifiers = mongoDBExportPartitionSupplier.apply(exportPartition);
+
+                    createDataQueryPartitions(exportPartition.getPartitionKey(), Instant.now(), partitionIdentifiers);
+                }
+                try {
+                    Thread.sleep(DEFAULT_TAKE_LEASE_INTERVAL_MILLIS);
+                } catch (final InterruptedException e) {
+                    LOG.info("The ExportScheduler was interrupted while waiting to retry, stopping processing");
+                    break;
+                }
+            } catch (final Exception e) {
+                LOG.error("Received an exception during export from DocumentDB, backing off and retrying", e);
+                try {
+                    Thread.sleep(DEFAULT_TAKE_LEASE_INTERVAL_MILLIS);
+                } catch (final InterruptedException ex) {
+                    LOG.info("The ExportScheduler was interrupted while waiting to retry, stopping processing");
+                    break;
+                }
+            }
+        }
+        LOG.warn("Export scheduler interrupted, looks like shutdown has triggered");
+        executor.shutdownNow();
+
+    }
+
+    private void createDataQueryPartitions(final String collection,
+                                           final Instant exportTime,
+                                           final List<PartitionIdentifier> partitionIdentifiers) {
+        final AtomicInteger totalQueries = new AtomicInteger();
```

Comment:
Why is this an `AtomicInteger`? It is a stack variable so it will not be shared across threads.

If this is to access in a lambda, you can use an array instead.

```
final int[] totalQueries = { 0 }
```

inside the lambda:

```
totalQueries[0]++
```

Reply:
The totalQueries is used inside lambda and it more clean that using array counter. AtomicInteger may have slight overhead compared to array manipulation. Is performance main concern when using AtomicInteger ?

Reply:
The compactness is nice, but it also sends a signal of concurrency. So it can be harder to understand.

Performance is a secondary concern. We can keep it this way.